### ISSUE TITLE
GraphQL roadmap

### ISSUE THREAD

**Issue by enisdenjo** [kamranahmedse/developer-roadmap#3027]
I have created a PR for a GraphQL roadmap teaser here: #3026.

As per the [contribution document's new roadmap section](https://github.com/kamranahmedse/developer-roadmap/blob/master/contributing.md#new-roadmaps), here's the textual roadmap representation:

_* I would create the actual roadmap myself, but I don't really know how to._

- Introduction to GraphQL
  - What is GraphQL?
  - What problems does GraphQL solve?
  - Thinking in graphs
  - On the frontend
  - On the backend
- Queries
  - What are queries?
  - Fields
  - Aliases
  - Arguments
  - Variables
    - Definitions
    - Defaults
  - Fragments
    - Using variables inside fragments
    - Inline Fragments
      - Meta fields
  - Directives
- Mutations
  - What are mutations?
  - Multiple fields in mutations
  - Operation Name
- Subscriptions
  - What are subscriptions?
  - Event based subscriptions
  - Live queries
  - `@defer` and `@stream` directives
- Schema
  - Type System
  - Fields
  - Scalars
    - Int
    - Float
    - String
    - Boolean
    - ID
  - Enums
  - Objects
  - Lists
  - Interfaces
  - Unions
  - Arguments
    - Input types
- Validation
- Execution
  - Root fields
  - Resolvers
    - Synchronous
    - Asynchronous
    - Scalar coercion
    - Lists
  - Validation
  - Producing the result
- Serving over internet
  - Over HTTP
    - GraphQL over HTTP specification
    - Caching
    - Batching
    - Authorization
  - Over WebSockets
    - GraphQL over WebSocket specification
    - Real-time
    - Authorization
  - Over SSE
    - GraphQL over Server-Sent Events specification
    - Authorization
- Pagination
- Backend implementations
  - JavaScript
    - graphql-http
    - GraphQL Yoga
    - Apollo Server
    - mercurius
- Frontend implementations
  - JavaScript
    - graphql-http
    - Relay
    - Apollo Client
    - Urql

I've noticed you guys are using [balsamiq](https://balsamiq.com) for the roadmaps, but how do you convert it to JSON? I am willing to help here too!

Furthermore, I want to write the content for the roadmap nodes - how do we make this connection?

Also, if the roadmap nodes change, how do we reflect that?

Looking forward to this 🚀!

**Comment by kamranahmedse** [comment]
Thank you for the contribution @enisdenjo -- the roadmap looks great!

As for the next steps, we will prepare and launch the interactive version of the roadmap this week. After that, you can help add the content to the roadmap nodes.

Regarding the changes to the roadmap, it's a bit tedious at the moment; you will have to open an issue on GitHub, and we will implement the changes. However, we are improving it in the next version, which is set to launch in February 2023, and will have the updated contribution guidelines for that.

**Comment by kamranahmedse** [comment]
@enisdenjo, we have implemented your roadmap at [roadmap.sh/graphql](https://roadmap.sh/graphql). Please let us know if you have any feedback or suggestions. 

Regarding the content contribution, you can submit a pull request by updating the [markdown files in the GraphQL roadmap directory](https://github.com/kamranahmedse/developer-roadmap/tree/master/content/roadmaps/116-graphql/content). Here are the [contribution guidelines if you need them](https://github.com/kamranahmedse/developer-roadmap/blob/master/contributing.md#adding-content). Let me know if you need any assistance.

Thank you again for your contribution!

**Comment by enisdenjo** [comment]
Hey @kamranahmedse, thank you for pushing the roadmap! Looks lovely.

One thing we've noticed is that the frontend<->backend implementations are inverted (see screenshot). Would be great if you could switch it around.

![image](https://user-images.githubusercontent.com/11807600/208478781-8485c7e5-428c-4d64-b339-ba1f37b002d1.png)

**Comment by kamranahmedse** [comment]
Fixed, thanks @enisdenjo

**Comment by meenakshi-dhanani** [comment]
Had some feedback - https://github.com/kamranahmedse/developer-roadmap/issues/3319

**Comment by kamranahmedse** [comment]
We have added the content [to the roadmap](https://roadmap.sh/graphql). Feel free to suggest any changes. Thank you!

**Comment by enisdenjo** [comment]
Closes #3319, thank you @meenakshi-dhanani for the excellent feedback!

@kamranahmedse here's the necessary changeset.

```diff
* Introduction to GraphQL
  * What is GraphQL?
  * What problems does GraphQL solve?
  * Thinking in graphs
  * On the frontend
  * On the backend
* Queries
  * What are queries?
  * Fields
  * Aliases
  * Arguments
  * Variables
    * Definitions
    * Defaults
  * Fragments
    * Using variables inside fragments
    * Inline Fragments
      * Meta fields
  * Directives
* Mutations
  * What are mutations?
  * Multiple fields in mutations
  * Operation Name
* Subscriptions
  * What are subscriptions?
  * Event based subscriptions
  * Live queries
  * `@defer` and `@stream` directives
* Schema
  * Type System
  * Fields
  * Scalars
    * Int
    * Float
    * String
    * Boolean
    * ID
  * Enums
  * Objects
  * Lists
  * Interfaces
  * Unions
  * Arguments
    * Input types
* Validation
* Execution
  * Root fields
  * Resolvers
    * Synchronous
    * Asynchronous
    * Scalar coercion
    * Lists
  * Validation
  * Producing the result
* Serving over internet
  * Over HTTP
    * GraphQL over HTTP specification
    * Caching
    * Batching
    * Authorization
  * Over WebSockets
    * GraphQL over WebSocket specification
    * Real-time
    * Authorization
  * Over SSE
    * GraphQL over Server-Sent Events specification
    * Authorization
* Pagination
-* Backend implementations
+* Implementations
  * JavaScript
+   * GraphQL.js
-   * graphql-http
-   * GraphQL Yoga
-   * Apollo Server
-   * mercurius
+ * Go
+   * GraphQL Go
+ * Java
+   * GraphQL Java
+* Servers
+ * JavaScript
+   * graphql-http
+   * GraphQL Yoga
+   * Apollo Server
+   * mercurius
-* Frontend implementations
+* Clients
  * JavaScript
    * graphql-http
    * Relay
    * Apollo Client
    * Urql
```

P.S. IDK whether you expect a separate issue for each suggestion or? Also is this diff format the easiest to interpret?
